Intersection

brettz9 opened the pull request:
Fixes part of #50.

feat: add intersection type (typescript and permissive only for now until jsdoc/closure may accept)

Note that I did not do an exhaustive analysis of where intersections may occur or what its operands may be, and merely copied the types for unions. (I would have thought that intersections might prevent string/number literals, but the TS playground does not seem to have an issue with it.)

sandersn commented:
> I would have thought that intersections might prevent string/number literals

Do you mean `1 & 2` or `"hello" & "goodbye"`? Vacuous intersections collapse to `never`, the bottom type, which is only usable to indicate errors. But intersecting two types is never an error; it may just create a type that's not practically usable.

sandersn commented on lib/publishing.js:
```diff
@@ -61,6 +61,10 @@ function createDefaultPublisher() {
       return format('%s|%s', concretePublish(unionNode.left),
                     concretePublish(unionNode.right));
     },
+    INTERSECTION (unionNode, concretePublish) {
```
I would prefer to format both intersections and unions with whitespace.

brettz9 replied:
Agreed. Added!

brettz9 commented:
Btw, I added a link to the TS intersection links, but note that the TS docs on JSDoc only explicitly mention support for unions, not intersections (they mention supporting any TS type, but it doesn't say any TS operator). However, it seems from comments others have made (and the issue requesting adding support), that they are supported within jsdoc.

I also added a link to the jsdoc issue filed that is requesting Closure support but surprisingly could find no issue requesting support for intersections in Closure (and Closure docs do not mention it as a supported operator and searching source seems to only turn up ampersands for the JS bitwise ampersand), so it seems reasonable to keep excluding it for closure mode.

sandersn commented:
> the TS docs on JSDoc only explicitly mention support for unions, not intersections (they mention supporting any TS type, but it doesn't say any TS operator)
I think that's a case of not explaining our terminology enough in the docs, then, because we consider type operators, like `&`, to create new type kinds, like intersections. The JSDoc page should probably link to the types/advanced types pages as well.

brettz9 commented:
Sure, yeah I figured it may be considering them as types, but a little thrown off with unions being specifically called out (and not seeming like just as an example).
